### ISSUE TITLE
Elliptic.verify verifies even when data is altered

### ISSUE THREAD

**Issue by sanandnarayan** [indutny/elliptic#143]
Hi, you can find in my code below that I altered data by appending 'p' in the front. But even then elliptic.verify returns true

```js
var EC = require('elliptic').ec;
var ec = new EC('secp256k1');

function Elliptic(){
}
Elliptic.verify = function(pubKey, msg, signature){
    var key = ec.keyFromPublic(pubKey,'hex');
    return key.verify(msg, signature);
};
Elliptic.sign = function(pvtKey, msg){
    var key = ec.keyFromPrivate(pvtKey);
    return key.sign(msg).toDER('hex');
};

var data = 'e6cb0a3dfd1e8850d0c6cbea768aded74fb690175738b49a3968cef3c7feaa38104022cd59f1aa916fa52edefa150cfffcfc64290f5694db801f120a977315b29eddde912bb316e7145f378843eaaf25c8ebb3b0f1ead690177ba3f835c100034ddcodeanand0.0015';
var sign = Elliptic.sign('905b6c7755f9354cf6977e4ca0e86f6f5874ed5e0a357ce9f418171d4f8d24f8', data);
var truth = Elliptic.verify('04022cd59f1aa916fa52edefa150cfffcfc64290f5694db801f120a977315b29eddde912bb316e7145f378843eaaf25c8ebb3b0f1ead690177ba3f835c100034dd', 'p'+data, sign);
```

**Comment by lacksfish** [comment]
Just curious, is the string `codeanand0.0015` at the end of `data` of any meaning?

**Comment by sanandnarayan** [comment]
@lacksfish yes.

**Comment by dcousens** [comment]
@sanandnarayan your `data` is expected to be `hex`. 
No doubt the `p` is ignored.

**Comment by sanandnarayan** [comment]
@dcousens thank you very much. 
What would you recommend for validating my data?

**Comment by aultac** [comment]
I'm having a very similar issue, and can replicate it on @sanandnarayan's repl even with all non-hex characters removed.  

Updated repl: https://repl.it/repls/WobblyLikableNet

Basically I took out all the non-hex characters, and then append a string of hex data to the end of the data to be verified and it still verifies.  It seems to be random: if I append `'1'` to the data then it correctly fails to verify, but if I add `'11'` to the data, then it incorrectly says the data matches the signature.  Seems to be related to the number of characters I add: adding an even number of characters results in improper verification, but adding an odd number of characters results in proper behavior.  Easy to play around with that repl and try the various options (look for the `data+'1'` in the verify line).

**Comment by dcousens** [comment]
@aultac but are you upholding the assertion `hex.length % 2 == 0`

**Comment by aultac** [comment]
nope, and that would be the problem :).  I fixed the problem in my other app by just sha256 hashing the message first as I should have in the first place.  I suspect that was @sanandnarayan 's problem too since his code was using the raw data as the message to sign.

**Comment by dcousens** [comment]
@aultac indeed,  the message should be an exact length of 32-bytes anyway IIRC.
